First draft workshop content (#3)
observability-lab-cse/observability-lab-workshop: 03-add-basic-observability-instrumentation/README.md

> Goal: Exploring different options to instrument applications with standard tooling and easily integrate them into Azure for data collection and observability.

<!--
In this section we are going to build the observability solution step by step by using automatic OpenTelemetry instrumentation.

> 1. OtelCollector introduction and ref to lower section with comparison to alternatives

Reviewer: ```suggestion
1. OpenTelemetry collector introduction and ref to lower section with comparison to alternatives
```
What do you mean with 'ref to lower section'?

2. Auto instrumentation of .Net application
3. Auto instrumentation of Java application

Additional topics:

- OtelCollector and the alternatives quick comparison. -->

<!-- Order of chapters and when to deploy otel collector -->

Now, let's dive into the exciting part of this workshop.

The goal of this section is to instrument the application we developed earlier to gain more visibility into its availability and health.

There are a few steps we need to take to achive this:

1. Provision necessary resources.
<!-- Consider provisioning manually, potentially. -->
2. Set up a mechanism to send application level telemetry data to these created resources.

> There are multiple ways to instrument your application, or even whole solution in more generally terms.

Reviewer: ```suggestion
There are multiple ways to instrument your application, or even whole solution in more general terms.
```

> Given we have an application running on AKS, which already provides workload level visibility regarding the health of the resources etc. However, what we're missing, as previously discussed, are application-level insights.

Reviewer: I think this section could be simplified a little:
```suggestion
Our application is running in AKS, which already provides workload level visibility regarding the health of the resources. However, what we're missing, as previously mentioned, are application-level insights.
```

> Note: The [section below](TODO) offers a comparison of different approaches for instrumenting applications, along with links to relevant resources and samples.

But first, let's provision our resources so that we have a destination to send the newly gathered data.

## Provision Resources

Can you guess the first resource we need for this section? If you said Application Insights, then ding, ding, ding... 100 points to you!

As expected, we require Application Insights. However, we'll also need a Log Analytics workspace to support our Application Insights instance.

> If you're curious about how Log Analytics and Application Insights work together, explore this section [Azure 101](tbd) for more information.

First, let's create a Log Analytics workspace in our resource group. Feel free to use either the Azure portal or any other tools for this task.

> Afterward, you can create an Application Insights resource, and be sure to specifically select the just-created Log Analytics workspace as the backing for the new Application Insights resource.

Reviewer: ```suggestion
Afterwards, you can create an Application Insights resource, and make sure to specifically select the just-created Log Analytics workspace as the backing instance for the new Application Insights resource.
```

## OpenTelemetry

> As mentioned there are many ways applications can be instrumented. Some require more custom code, some less.

Reviewer: ```suggestion
As mentioned, there are many ways to instrument applications. Some require writing more custom code, some less.
```

For this workshop, let's use an approach that requires no changes to the application code, as it will come in handy when working with preexisting applications.
To do so, we will make use of the OpenTelemetry Auto instrumentation for different programming languages and the [Otel Collector](TODO).

### Configure Applications

> For other auto-instrumentation, our applications need a small "agent" running alongside

Reviewer: ```suggestion
For auto-instrumentation, our applications need a small "agent" running alongside
```

the service to be able to gather telemetry data from common logging, metric, and tracing libraries.

> Let's have a look at the applications. Here is a link to the [OpenTelemetry Auto-instrumentation for Java](https://github.com/open-telemetry/opentelemetry-java-instrumentation) and [OpenTelemetry Auto-instrumentation for C#](). Try to adjust the application's Docker file and configs in such a way that we can scrape the logs and send them to the OpenTelemetry collector.

Reviewer: ```suggestion
Let's have a look at the applications. Here is a link to the [OpenTelemetry Auto-instrumentation for Java](https://github.com/open-telemetry/opentelemetry-java-instrumentation) and [OpenTelemetry Auto-instrumentation for C#](https://opentelemetry.io/docs/instrumentation/net/automatic/). Try to adjust the application's Docker file and configs in such a way that we can scrape the logs and send them to the OpenTelemetry collector.
```

RUN ./gradlew build

ENTRYPOINT ["java", "-javaagent:opentelemetry-javaagent.jar", "-jar","build/libs/devices-api.jar"]
```

</details>

The same again for the C# service.

<details markdown="1">
<summary>Click here for the Dockerfile with the auto-instrumentation Agent for C#</summary>

```Docker
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
WORKDIR /src
COPY ["DeviceManager.csproj", "."]
RUN dotnet restore "./DeviceManager.csproj"
COPY . .
WORKDIR "/src/."
RUN dotnet build "DeviceManager.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "DeviceManager.csproj" -c Release -o /app/publish /p:UseAppHost=false

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .

ARG OTEL_VERSION=1.0.0-rc.2
ADD https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/releases/download/v${OTEL_VERSION}/otel-dotnet-auto-install.sh otel-dotnet-auto-install.sh
RUN apt-get update && apt-get install -y unzip && apt-get install -y curl && \
    OTEL_DOTNET_AUTO_HOME="/otel-dotnet-auto" sh otel-dotnet-auto-install.sh

ENTRYPOINT ["dotnet", "DeviceManager.dll"]
```

</details>

> Nope, let's build the images with the agent and add a new tag to it, so we know which images are auto-instrumented. TODO

Reviewer: ```suggestion
Now, let's build the images with the agent and add a new tag to it, so that we know which images are auto-instrumented. TODO
```
